refactor tracing so that it's straightforward to run a callable with some specified trace

bavardage opened the pull request:
Usecase (there may be an existing way to do this)

I have a scheduled background task which makes requests to other services.

I want each run of this background task to run with a new trace id
- all calls to remote services as part of this single run have this new trace id in common
- all log statements made as part of this single run have the this new trace id in common

currently I do this via the somewhat janky

```java
void run() {
    String traceId = Tracers.randomId();
    Tracer.initTrace(Optional.empty(), traceId);
    MDC.put("traceId", traceId);
    // work of the scheduled task goes here
}
```

I should also probably remove the MDC when done (i.e. set it back to what it was before) and restore the traceid too, given potentially other scheduled tasks could in theory share this same scheduled executor service.

whereas with this PR I can do, I think
```java
void run() {
    Tracer.withTrace(Tracers.createTrace(Optional.empty(), () -> {
        //work of the scheduled task goes here
    });
}
```

Happy to consider other approaches - perhaps there exists something - or maybe this looks like a wrapper around a scheduled executor service, so that each scheduled call is automatically wrapped in a new trace.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/http-remoting/472)
<!-- Reviewable:end -->

Commit:
refactor tracing so that it's straightforward to run a callable with some specified trace

bavardage commented:
(also I didn't write tests yet, wanted to get thoughts on approaches first..)

uschi2000 commented:
can you explain the use-case a little more? what is there an existing trace that you don't want to push down into executors/callables?

bavardage commented:
I'm not sure I understand your last question.

I'm running a scheduled background task via a [ScheduledExecutorService](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html)

There's no request involved (i.e. the [TraceEnrichingFilter](https://github.com/palantir/http-remoting/blob/develop/jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/TraceEnrichingFilter.java) will not have run).

The only way an existing trace would be present would be by accident/pollution from some other task sharing the same executor service. I.e, without this, the behaviour is

 - the first remote call made by the scheduled task will result in the (thread local) trace id being set
 - that happens though, I think, via the thread local init, so this trace id isn't set in the mdc
    - ...and so no log statements made will actually get that trace id
    - you'll only really see it showing up in the request logs of the services the task is calling
 - all subsequent runs of the task that happen to re-use this thread will now get the same traceid
 - traceid ceases to be meaningful

maybe easier to explain in person?

markelliot commented:
What about Tracers.wrap(Callable) and Tracers.wrap(Runnable)?

On Jul 9, 2017, at 22:18, Ben Duffield <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm not sure I understand your last question.

I'm running a scheduled background task via a ScheduledExecutorService[docs.oracle.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__docs.oracle.com_javase_7_docs_api_java_util_concurrent_ScheduledExecutorService.html&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=GRXZQXkMkJg9eOWsw9WwtA5TjfOpcymgXUpTWrBVLRw&s=oUGq2edbdM1itknrG6smVHULDIRBcVMH6UTdYPFtTr0&e=>

There's no request involved (i.e. the TraceEnrichingFilter[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_http-2Dremoting_blob_develop_jersey-2Dservers_src_main_java_com_palantir_remoting2_servers_jersey_TraceEnrichingFilter.java&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=GRXZQXkMkJg9eOWsw9WwtA5TjfOpcymgXUpTWrBVLRw&s=zOy-l-YMji74NZpSr6Cdx_QCu6L7TMaEDIPBtI3E5CY&e=> will not have run).

The only way an existing trace would be present would be by accident/pollution from some other task sharing the same executor service. I.e, without this, the behaviour is

  *   the first remote call made by the scheduled task will result in the (thread local) trace id being set
  *   that happens though, I think, via the thread local init, so this trace id isn't set in the mdc
     *   ...and so no log statements made will actually get that trace id
     *   you'll only really see it showing up in the request logs of the services the task is calling
  *   all subsequent runs of the task that happen to re-use this thread will now get the same traceid
  *   traceid ceases to be meaningful

maybe easier to explain in person?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_http-2Dremoting_pull_472-23issuecomment-2D313959622&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=GRXZQXkMkJg9eOWsw9WwtA5TjfOpcymgXUpTWrBVLRw&s=Cs-Q6EHahMYKPdQnWcscNqaWU98wj2cphMPZsMaMGnw&e=>, or mute the thread[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAp3QwFIzg-2Dpx-2DkbfLC3gBLYq8MBXK2hks5sMUOFgaJpZM4ORejq&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=GRXZQXkMkJg9eOWsw9WwtA5TjfOpcymgXUpTWrBVLRw&s=Z6d_TaeWKEimrN-7oarKka1hEQHcQCWy83-vYHLzrU8&e=>.

bavardage commented:
What about them?

Their behaviour is they make sure that when the callable/runnable actually runs, it runs with the traceid present at the time of construction of the callable/runnable.

This is a scheduled task - constructed at server start - and thus there is no trace id around when it is constructed.

This is sorta the opposite of `Tracers#wrap`

markelliot commented:
Got it. I'd think that'd work best with a wrapWithoutTrace rather than exposing the withTrace method.

On Jul 9, 2017, at 22:26, Ben Duffield <notifications@github.com<mailto:notifications@github.com>> wrote:


What about them?

Their behaviour is they make sure that when the callable/runnable actually runs, it runs with the traceid present at the time of construction of the callable/runnable.

This is a scheduled task - constructed at server start - and thus there is no trace id around when it is constructed.

This is sorta the opposite of Tracers#wrap

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_http-2Dremoting_pull_472-23issuecomment-2D313961944&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=DY_JXyrmx6_MLR4aIQA3ZcvoqrdVoD4arx0DTa_0yrs&s=9WRysO2an90lbqhZaMaPTBoCOS9F5igy6Ojk6JS5YZk&e=>, or mute the thread[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAp3Q5BDVJUjwp7HQLcsVAh8JtXqcHl0ks5sMUVngaJpZM4ORejq&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=DY_JXyrmx6_MLR4aIQA3ZcvoqrdVoD4arx0DTa_0yrs&s=aWcqNEwNfvtP1lTU_KFbXdKqAoPZvjbVxTPQ31waEt4&e=>.

bavardage commented:
where `wrapWithoutTrace` takes care of making a new traceid? 

I was also pondering whether what I actually want here is a `TracingAwareScheduledExecutorService` which 'magically' handles all of this for you

markelliot commented:
There is one of those but it ultimately calls wrap, and you want not that.

On Jul 9, 2017, at 22:49, Ben Duffield <notifications@github.com<mailto:notifications@github.com>> wrote:


where wrapWithoutTrace takes care of making a new traceid?

I was also pondering whether what I actually want here is a TracingAwareScheduledExecutorService which 'magically' handles all of this for you

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_http-2Dremoting_pull_472-23issuecomment-2D313967777&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=IkbBS3hLSxXFS-pV4-yl6GRT38vCxK4lcLBjdYJxeNQ&s=67kUvKoYKFHnx3jpFh142aDVKje6O8fmWkB-nWvgy0Q&e=>, or mute the thread[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAp3Qxyp9NlrBqXtmdfZnTQKSpbCYdgYks5sMUr4gaJpZM4ORejq&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=Vp81aOxWZvuVgVK_wp-VF3pIYG92B19LcCw6XKeYC0U&m=IkbBS3hLSxXFS-pV4-yl6GRT38vCxK4lcLBjdYJxeNQ&s=d0GA1fw4kh40w9zB_XzP2vJ-7l-qnhrE_uqgYD90KX8&e=>.

bavardage commented:
nope, there isn't for a *Scheduled*ExecutorService

Commit:
expose wrapWithoutTrace

Commit:
docs

bavardage commented:
made the changes/refactored this on top of the for-3.0.0 branch

markelliot commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracer.java:
```diff
@@ -51,14 +51,23 @@ protected Trace initialValue() {
+     * Trace#isObservable observable} iff the given flag is true, or, iff {@code isObservable} is absent,
+     * if the {@link #setSampler configured sampler} returns true.
+     */
+    public static Trace createTrace(Optional<Boolean> isObservable, String traceId) {
```
I don't think you need this or at least don't need it to be exposed as public anymore.

markelliot replied:
Also instead of "isObservable" it'd be clearer to use a Boolean "useSampler", or perhaps, since we always want to use the sampler, remove thenparameter?

markelliot replied:
And maybe even remove the traceId parameter. Ideal might be we add a clearTrace method which returns a copy of the current trace and zeros the current state, then use that method in "withTrace" instead of copy/set. At that point, this method goes away.

bavardage replied:
I use `createTrace` in `Tracers#wrapWithoutTrace`

markelliot replied:
Yes, and it could be package private, or per my 3rd comment, be removed in favor of a clearer purpose "clearTrace" method.

bavardage replied:
made it package. I don't really understand how one would use `clearTrace` as you describe - it seems you'd have to call `initTrace` immediately before, which seems a fairly roundabout way of getting a fresh trace.

bavardage replied:
#473 for FLUP

markelliot commented:
We have a WrappingScheduledExecutorService and one of the Tracers#wrap methods will configure it for many of the calls. If you want to edit which methods we wrap which way, feel free.

bavardage commented:
Will look at changing the `WrappingScheduledExecutorService` in FLUP

Commit:
-> package private

Commit:
fix wrapping behaviour

markelliot commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracer.java:
```diff
@@ -51,14 +51,23 @@ protected Trace initialValue() {
+     * Trace#isObservable observable} iff the given flag is true, or, iff {@code isObservable} is absent,
+     * if the {@link #setSampler configured sampler} returns true.
+     */
+    /*package*/ static Trace createTrace(Optional<Boolean> isObservable, String traceId) {
```
+ remove params?

Commit:
some more refactoring

- set the trace id in the MDC when setting the current trace
- add a method `Tracer#getAndClearTrace`
- use `getAndTraceTrace` for `runWithoutTrace`

markelliot commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracer.java:
```diff
@@ -213,6 +221,9 @@ static Trace copyTrace() {
         currentTrace.set(trace);
+
+        // Give SLF4J appenders access to the trace id
+        MDC.put(Tracers.TRACE_ID_KEY, trace.getTraceId());
```
we should do this in createTrace as well?

markelliot commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracer.java:
```diff
@@ -195,6 +196,13 @@ public static String getTraceId() {
+    /** Clears the current trace id and (a copy of) it. */
+    public static Trace getAndClearTrace() {
+        Trace trace = currentTrace.get();
+        currentTrace.remove();
```
\+ unset MDC?

markelliot commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracers.java:
```diff
@@ -86,26 +85,27 @@ public static Runnable wrap(Runnable delegate) {
      * will have a fresh trace.
      */
     public static <V> Callable<V> wrapWithoutTrace(Callable<V> delegate) {
```
sorry to nit pick: `wrapWithNewTrace`

markelliot commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracers.java:
```diff
@@ -79,6 +79,34 @@ public static Runnable wrap(Runnable delegate) {
+     * Runs the given callable after firstly clearing the current trace. The current
+     * trace is restored after running the callable.
+     */
+    private static <T> T runWithoutTrace(Callable<T> callable) throws Exception {
```
maybe just inline this?

Commit:
CR feedback

Commit:
fix typo

uschi2000 commented on jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/TraceEnrichingFilter.java:
```diff
@@ -59,10 +58,6 @@ public void filter(ContainerRequestContext requestContext) throws IOException {
             }
         }
 
-        // Give SLF4J appenders access to the trace id
```
hopefully people upgrade jersey-servers and tracers at the same time ...

bavardage replied:
are we doing `com.palantir.remoting3` for remoting3?

bavardage replied:
... I guess, not for everything

uschi2000 commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracer.java:
```diff
@@ -186,6 +198,14 @@ public static String getTraceId() {
         return currentTrace.get().getTraceId();
     }
 
+    /** Clears the current trace id and (a copy of) it. */
```
+"returns"

bavardage replied:
done

Commit:
javadoc typo

uschi2000 commented on tracing/src/main/java/com/palantir/remoting2/tracing/Tracers.java:
```diff
@@ -79,6 +79,28 @@ public static Runnable wrap(Runnable delegate) {
+     * construction or any trace already set on the thread used to execute the callable. Each execution of the callable
+     * will have a fresh trace.
+     */
+    public static <V> Callable<V> wrapWithNewTrace(Callable<V> delegate) {
```
should have some tests.

uschi2000 reviewed (commented):
looks good, but would like to see test coverage for the new functionality.

Commit:
test wrap callable

Commit:
test for mdc setting, clearing

Commit:
trace construction test

Commit:
test state restored

uschi2000 commented:
more checkstyle, sorry. otherwise lgtm.

Commit:
test class final
